인벤토리 데이터구조 개편, 아이템 창 구현

adlet0331 opened the pull request:
Intentory 데이터구조 개편(Dictionary -> List Array
아이템을 보여주는 ItemWindow와 InventoryWindow 분리 및 완성. (드래그해서 위치 바꾸는 기능 + 가방에 넣는 기능)
탐험 아이템 선택 창, 간단한 월드맵(기능 구현 X) 구현

adlet0331 committed:
ItemWindow 제작중

Co-Authored-By: Adlet <48337548+adlet0331@users.noreply.github.com>

adlet0331 committed:
탐험 인벤토리와 인벤토리 창 동기화 작업중

탐험 인벤토리와 인벤토리 창 동기화 작업중

adlet0331 committed:
ItemWindow 동기화작업 완료, Draggable 기능 구현 시작

ItemWindow 동기화작업 완료, Draggable 기능 구현 시작

adlet0331 committed:
인벤토리,가방 간 상호작용 가능

인벤토리,가방 간 상호작용 가능

adlet0331 committed:
드래그 Ondrop 제외 구현 완료 / Panel 버그 수정

드래그 Ondrop 제외 구현 완료 / Panel 버그 수정

adlet0331 committed:
ItemIcon Drag 기능 추가(같은공간 안에서만)

ItemIcon Drag 기능 추가(같은공간 안에서만)

adlet0331 committed:
WorldMapWindow 추가

WorldMapWindow 추가

Signed-off-by: Adlet <eypk777@gmail.com>

adlet0331 committed:
Merge remote-tracking branch 'origin/master' into Adlet

Signed-off-by: Adlet <eypk777@gmail.com>

# Conflicts:
#	Assets/Scenes/TestTownRoamStartScene_Adlet.unity
#	Assets/Scripts/Data/Inventory.cs
#	Assets/Scripts/GameInput.cs
#	Assets/Scripts/Windows/ExplorationCharacterSelectingWindow.cs
#	Assets/Scripts/Windows/ExplorationItemSelectingWindow.cs
#	Assets/Scripts/Windows/UiWindowsManager.cs

Jaewookim08 commented on Assets/Scripts/Data/Inventory.cs:
```diff
@@ -1,38 +1,179 @@
+            item = _item;
+            itemnum = _itemnum;
+        }
+        public void setItemStack(Item _item, int _itemnum)
+        {
+            item = _item;
+            itemnum = _itemnum;
+        }
+        public (Item, int) getItemStack()
+        {
+            return (item, itemnum);
+        }
```
그냥 Item과 ItemNum 변수를 public으로 푸는 것과 차이가 없는 것 같습니다.

Jaewookim08 commented on Assets/Scripts/Data/Inventory.cs:
```diff
@@ -1,38 +1,179 @@
-        
+        public List<ItemStack>[] InventoryAList => inventoryAList;
+        public int WindowSize => windowsize;
+        public int TotalSize => totalsize;
```
Auto-Implemented Properties을 사용하면 더 좋을 것 같습니다.

Jaewookim08 commented on Assets/Scripts/Data/Inventory.cs:
```diff
@@ -1,38 +1,179 @@
+        }
+        private Item item;
+        private int itemnum;
+    }
```
그냥 Item과 ItemNum 변수를 public으로 푸는 것과 차이가 없는 것 같습니다.

Jaewookim08 commented on Assets/Scripts/Data/Inventory.cs:
```diff
@@ -1,38 +1,179 @@
+    }
     public class Inventory
     {
+        public Inventory(int _WindowSize, int _TotalSize, bool _IsStacked)
```
클래스 내부의 변수명과 겹치는 게 문제라면 this. 으로 구분할 수 있습니다.

Jaewookim08 commented on Assets/Scripts/GameInput.cs:
```diff
@@ -14,16 +15,45 @@ private void Update()
+                else if (UiWindowsManager.Instance.TextureWindow.IsOpen)
                 {
                     UiWindowsManager.Instance.TextureWindow.Toggle();
                 }
```
UiWindowsManager.CloseSingleWindow 같은 함수 만들어서 거기로 옮겨주세요. 윈도우가 닫히는 순서는 그쪽 책임입니다.

Jaewookim08 commented on Assets/Scripts/GameInput.cs:
```diff
@@ -14,16 +15,45 @@ private void Update()
-            if (Input.GetKeyDown(KeyCode.I)) //인벤토리
+            if (Input.GetKeyDown(KeyCode.I)) // 인벤토리
+            {
+                UiWindowsManager.Instance.itemInfoPanel.Close(ItemIcon.panelRequestId); // 마우스 올리고 I 누르면 pannel이 안 꺼져서 추가함. 아래도 동일한 이유
```
InventoryWindow가 처리해야 하는 내용. Issues에 추가하면 수정하겠습니다. GameInput이 해결해야 하는 문제는 아니므로 원래대로 돌려주세요.

Jaewookim08 commented:
ExplorationCharacterSelectingWindow.cs.meta랑 ExplorationItemSelectingWindow.cs.meta가 깃에 포함 안된듯

Jaewookim08 commented on Assets/Scripts/Windows/Inventory/InventoryWindow.cs:
```diff
@@ -94,49 +58,11 @@ public void ChangeCategory(ItemCategory category)
+            itemWindow.RefreshCategory(category);
             Refresh();
             _changeCategoryBlock = false;
         }
```
ChangeCategory 왜 이렇게 코드를 이상하게 짰나 하고 보니까 제 코드네요.
수정하겠습니다....

Jaewookim08 commented on Assets/Scripts/Windows/ItemWindow.cs:
```diff
@@ -0,0 +1,116 @@
+                    {("확인", () => ItemClick(index)), ("취소", () => { }) });
+        return;
+    }
+    private void InitiateIcons()
```
InitializeIcons를 잘못 쓴 듯?

Jaewookim08 commented on Assets/Scripts/Windows/ItemWindow.cs:
```diff
@@ -0,0 +1,116 @@
+        for (int i = 0; i < _inventory.WindowSize; i++)
+        {
+            _ItemIconList.Add(Instantiate(ItemIconPrefab, transform));
+            _ItemIconList[i].UpdateAppearance(null, 0, i, _inventory, this);
```
Instantiate의 반환값을 이용해 UpdateAppearance를 부르는 게 낫습니다. 방금 생성한 itemIcon을 i번째 요소가 일 것이라고 생각하고 진행하는 것은 이 경우엔 틀리진 않았습니다만 에러가 나기 쉽고 보기에 직관적이지도 않습니다. 아니면 아예 전부 추가한 후에 밑에 foreach를 하나 더 넣어 UpdateAppearance를 부르는 것도 좋습니다.

adlet0331 replied:
foreach문을 추가하겠습니다.

adlet0331 committed:
실행 오류 수정, PR-두 UI Window 깃에 추가

타운 -> 탐색 함수 WorldMapWindow에 추가 필요

adlet0331 committed:
실행 오류 수정, PR-두 UI Window 깃에 추가

실행 오류 수정, PR-두 UI Window 깃에 추가

adlet0331 committed:
Merge branch 'Adlet' of https://github.com/honeyirene/guildmaster into Adlet

Jaewookim08 commented on Assets/Scripts/Data/Inventory.cs:
```diff
@@ -1,38 +1,179 @@
+    }
     public class Inventory
     {
+        public Inventory(int _WindowSize, int _TotalSize, bool _IsStacked)
```
Inventory는 이상한 점이 좀 많습니다. 일단 WindowSize와 TotalSize를 받는다는 게 조금은 이상하게 느껴집니다 . _TotalSize를 WindowSize로 나누어서 category 개수가 자유롭게 결정되는데(_TotalSize가 WindowSize의 배수가 아니면 _TotalSize가 total size가 아니게 되기도 하고요), 밑에 getItemToCategoryNum을 보면 category는 제한되어 있어야 합니다. AddItem에서는 카테고리 분류 기능이 Inventory에 있지만 ChangeItemIndex는 caller쪽에서 아이템의 카테고리를 구한 뒤에 불러야 한다는 것도 이상합니다. 그러니까 코드 일부는 카테고리가 딱 일정 종류에 한정되어 있고 인벤토리가 알아서 처리해 줄 것처럼 되어 있으면서 일부는 category가 그냥 int 사이즈 값인 것처럼 짜여 있습니다. 인벤토리가 꼭 여러 카테고리를 다 담아야 할까요? 카테고리 관련 코드를 빼면 훨씬 클래스가 깔끔해질 것 같습니다.

honeyirene replied:
Inventory 를 Inventory랑 Inventroy entry 등 좀 개념상 나누는게 좋을거같고 size관련 문제는 window 즉 UI 쪽 코드에서 처리하는게 맞을거같아요 
Inventory를 아이템을 받고 넘기고, 어떤 템을 가지고 있는지, 최대치에 걸리지는 않는지 등등. 데이터 관리를 핵심으로 작성하고 그에 필요한 부가 기능들을 다른 클래스 (Inventory entry나 InventoryWindow 등등)로 쪼개는게 맞을거같네요

adlet0331 replied:
인벤토리 하나를 상정하고 만든 코드라, 탐색에 가져갈 Bag같은 곳에서는 편법을 사용하여 구현한 점이 있습니다. (무조건 카테고리를 0으로 설정)

adlet0331 replied:
생각해보니 카테고리를 빼고 Inventory 위에 PlayerInventory를 만드는게 훨씬 괜찮을 것 같네요. 일단 그렇게 방향을 잡고 고쳐보겠습니다.
인벤토리를 하나에 담으려는 프레임에 갇혀있었던 것 같습니다.

honeyirene commented on Assets/Scripts/Data/Inventory.cs:
```diff
@@ -1,38 +1,179 @@
+    {
+        public Item Item => item;
+        public int ItemNum => itemnum;
+        public ItemStack(Item _item, int _itemnum)
```
ItemStack이 사실상 Inventory class 에서만 쓰이는데
비슷한 구조 다른 쪽에서 만들게 된다면 메인 class 안에 넣어서 private으로 바꾸는게 좋을거 같아요

honeyirene commented on Assets/Scripts/GameInput.cs:
```diff
@@ -14,16 +15,47 @@ private void Update()
+            }
             if (Input.GetKeyDown(KeyCode.P)) //캐릭터
                 UiWindowsManager.Instance.characterInspectWindow.Toggle();
             if (Input.GetKeyDown(KeyCode.G)) //길드 정보
                 UiWindowsManager.Instance .guildInspectWindow.Toggle();
+            if (Input.GetKeyDown(KeyCode.E))
+                UiWindowsManager.Instance.ExplorationCharacterSelectingWindow.Toggle();
             //Settings 와 그 자식 파일들 
```
위에 처럼 exploration 창 여는 키라고 주석 넣어주세요 👍

Jaewookim08 commented on Assets/Scripts/Windows/ExplorationCharacterSelectingWindow.cs:
```diff
@@ -0,0 +1,144 @@
+        public void Open()
+        {
+            base.OpenWindow();
+            if(first) Set_allCharacters();
```
첫 번째 열릴 때만 allCharacters를 설정하는 것 같은데 그러면 안 되지 않나 생각합니다.

adlet0331 replied:
allCharacters, exploreCharacters 둘 다 public으로 수정하고 Awake에서 allCharacters를 초기화하도록 하겠습니다.

adlet0331 replied:
그나저나 마을->탐색->마을로 씬이동이 된다면 2번째 마을 씬으로 갈 때 Awake가 또 불러와지나요? 그렇게 되면 탐색씬에서 탐색을 간 길드원들을 넘겨받아야 할 것 같네요.

adlet0331 replied:
네 파괴가 되었다가 다시 생성된다고 하네요.

Jaewookim08 commented on Assets/Scripts/Windows/ExplorationCharacterSelectingWindow.cs:
```diff
@@ -0,0 +1,144 @@
+            base.Close();
+            UiWindowsManager.Instance.ExplorationItemSelectingWindow.Open();
+            //ExplorationLoader.Instance.Load(_exploreCharacters); //Asd가 구현중인 기능 에러 뜨는게 정상
+        }
```
이 클래스가 다음 행동을 결정하지 말고, 선택한 캐릭터들을 반환하는 것 까지를 역할로 하면 더 좋을 것 같습니다.
그나저나 선택한 캐릭터들은 외부에서 어떻게 가져오는 거죠?

adlet0331 replied:
리스트를 public으로 할 생각입니다.

Jaewookim08 commented on Assets/Scripts/Windows/ExplorationItemSelectingWindow.cs:
```diff
@@ -0,0 +1,84 @@
+            Refresh();
+        }
+
+        public void GoWorldMap()
+        {
+            base.Close();
+            UiWindowsManager.Instance.worldMapWindow.Open();
+        }
+        public void Back()
+        {
+            base.Close();
+            UiWindowsManager.Instance.ExplorationCharacterSelectingWindow.Open();
+        }
```
마찬가지로, 다음/뒤로가기를 얘가 직접 처리하는 것보다 유저가 어떤 선택을 했는지와 선택한 아이템들을 반환하는 것까지를 역할로 하면 좋을 것 같습니다.

adlet0331 replied:
위와 마찬가지로 리스트를 Public으로 선언을 하겠습니다.

Jaewookim08 commented on Assets/Scripts/Windows/Inventory/ItemIcon.cs:
```diff
@@ -32,17 +53,63 @@ public void OnPointerExit(PointerEventData eventData)
+
+        protected override int EventArgument => index;
+
+        private static int _panelRequestId;
```
static이면 안 되는 변수인데 어째서 변경했나요?

Jaewookim08 commented on Assets/Scripts/Windows/Inventory/ItemIcon.cs:
```diff
@@ -32,17 +53,63 @@ public void OnPointerExit(PointerEventData eventData)
+            //categoryNum = Data.Inventory.getItemToCategoryNum(ItemWindow.DraggingItem.item);
+            //if (ItemWindow.DraggingItem.currentInventory.IsStacked) categoryNum = 0;
+            //var (_item2, _itemnum2) = ItemWindow.DraggingItem.currentInventory.InventoryAList[categoryNum][ItemWindow.DraggingItem.index].getItemStack();
+            //���� - �κ��丮 �� ��ȣ�ۿ�� ��� ¥��� �ߴ� ����. �ϴ� �κ��丮 - �κ��丮 or ���� - ���� ��ȣ�ۿ븸 ������
```
인코딩이 utf-8이 아닌 것으로 보입니다.

Jaewookim08 commented on Assets/Scripts/Windows/Inventory/ItemIcon.cs:
```diff
@@ -7,16 +8,36 @@
 namespace GuildMaster.Windows.Inventory
 {
-    public class ItemIcon: GenericButton<Item>, IPointerEnterHandler, IPointerExitHandler
+    public class ItemIcon : GenericButton<int>, IPointerEnterHandler, IPointerExitHandler, IDragHandler, IBeginDragHandler, IEndDragHandler, IDropHandler
```
부속품으로 쓰이는 오브젝트인데, 너무 맡는 역할이 많습니다.(재사용을 힘들게 합니다. ItemIcon을 사용하는 모든 상황이 이 기능들을 다 사용한다는 보장이 없으니까요:) ) 사실 그렇게 따지면 원래 제가 짠 패널 띄우는 코드도 여기에 있으면 안되긴 하네요. 여하튼 활용성을 높이려면 최대한 객체의 책임을 간단하게 하는 것이 좋습니다.

Jaewookim08 commented on Assets/Scripts/Windows/Inventory/ItemIcon.cs:
```diff
@@ -7,16 +8,36 @@
+        public static int panelRequestId => _panelRequestId;
 
-        public void UpdateAppearance(Item item, int number)
+        public void UpdateAppearance(Item item, int number, int index, Data.Inventory inventory, ItemWindow itemWindow)
```
ItemWindow에 대한 의존성은 넘어간다 치더라도 Inventory에 의존하는 건 문제가 좀 있습니다.

adlet0331 replied:
Drag Drop 기능 구현하면서 추가한 코드입니다. 
인벤토리에서 가방으로 옮기는 기능을 추가하려 추가했었는데, 이건 클릭 이벤트로 대체되어서 inventory 의존성 부분은 완전히 삭제하겠습니다.

Jaewookim08 commented on Assets/Scripts/Windows/Inventory/ItemIcon.cs:
```diff
@@ -32,17 +53,63 @@ public void OnPointerExit(PointerEventData eventData)
+        {
+            if (_item == null) return;
+            int _currentcategory = Data.Inventory.getItemToCategoryNum(_item);
+            ItemWindow.DraggingItem = this;
```
이 정도로 구현 깊숙이 있는 클래스에서 ItemWindow의 누구나 변경 가능한 static 변수를 조작하는 건 상당히 예측하기 어렵고 에러나기 쉬운 동작입니다.

adlet0331 replied:
새로운 클래스를 둬서 싱글톤화 시키고, public 변수로 선언하도록 변경해보겠습니다.

Jaewookim08 commented on Assets/Scripts/Windows/ItemWindow.cs:
```diff
@@ -0,0 +1,116 @@
+    [SerializeField] private ItemIcon ItemIconPrefab;
+    [SerializeField] private bool ClickEventOn = true;
+
+    public static ItemIcon DraggingItem;
```
public static 변수를 최대한 피합시다.

adlet0331 replied:
이것 또한 새로운 클래스를 둬서 싱글톤화 시키고, public 변수로 선언하도록 변경해보겠습니다.

Jaewookim08 commented on Assets/Scripts/Windows/ItemWindow.cs:
```diff
@@ -0,0 +1,116 @@
+using System.Linq;
+using UnityEngine;
+
+public class ItemWindow : MonoBehaviour
```
ItemIcon과 똑같이, 너무나 많은 역할을 맡은 클래스입니다.(재사용이 사실상 불가능해졌습니다) 사실 SetInventory(inventory) 함수로써 보여주는 아이템을 결정하는데 동작은 Player.Instance.Inventorys나 ...ExplorationItemSelectingWindow.ExploreInventory를 접근해서 한다는 것부터 무언가 심각하게 문제가 있습니다. 역할을 인벤토리를 보여주는 것 정도로 한정시켜 놓고 OnClick 등은 이벤트를 통해 위(클래스의 사용자)로 올려보내면 어떨까요.

adlet0331 replied:
그렇게 변경해보겠습니다.

Jaewookim08 reviewed (changes requested):
잘 짰다고 생각합니다. 제가 좀 깐깐하게 코멘트를 달긴 했는데, 대부분 큰 문제가 있는 거라기보다는 이렇게 하면 더 깔끔하지 않을까- 에 가까운 내용입니다. 다만 역할 구분이 불명확하거나 맡은 책임이 너무 많은 객체들이 많이 보입니다. 조금 더 객체들의 책임과 의존성에 신경 써서 고쳐 주었으면 합니다.

adlet0331 committed:
PR 수정

ItemStack 구조 수정, 탐색으로 넘어가는것 추가

adlet0331 reviewed (commented):
구현 전 피드백

adlet0331 committed:
PR 중

PlayerInventory구현, ItemIcon, Inventory 수정, ItemWindow => ItemListView 이름 변경, 수정

Signed-off-by: Adlet <eypk777@gmail.com>

adlet0331 committed:
[PR] Event 받는 것 이외에 모두 해결

버그 픽스 필요

adlet0331 reviewed (commented):
에러 고치는 것과 이벤트 처리만 남았습니다

adlet0331 committed:
Enum int 로 캐스팅

PR중

adlet0331 committed:
InventoryWindow Bug Fix

PR  중

Signed-off-by: Adlet <eypk777@gmail.com>

adlet0331 committed:
InventoryWindow Drag event 이외 수정 완료

PR중

Signed-off-by: Adlet <eypk777@gmail.com>

adlet0331 reviewed (commented):
일단은 완료했어요.

이벤트 구독이 겹치는 문제만 해결하면 진짜 끝.

adlet0331 committed:
이벤트 구독 겹침 수정필요

PR

honeyirene commented on Assets/Scripts/Data/Player.cs:
```diff
@@ -24,7 +24,7 @@ public static Player Instance
         // 속한 데이터
         public readonly QuestManager QuestManager;
-        public readonly Inventory Inventory;
+        public readonly PlayerInventory PlayerInventory;
```
이름 고민해봅시다

honeyirene commented on Assets/Scripts/Data/Inventory.cs:
```diff
@@ -1,38 +1,168 @@
+using System.Diagnostics;
+using UnityEngine;
 
 namespace GuildMaster.Data
```
namespace 연습

honeyirene commented on Assets/Scripts/Data/Inventory.cs:
```diff
@@ -1,38 +1,168 @@
+            Item = item;
+            ItemNum = itemNum;
+        }
+        public void setItemStack(Item item, int itemNum)
+        {
+            if (item == (Item)null && itemNum != 0) throw new ArgumentException("������ ���� null�Դϴ�");
+            Item = item;
+            ItemNum = itemNum;
+        }
```
이 함수를 inventory로 옮기고 정리해주세요

honeyirene replied:
```
        private void setItemStack(Item item, int itemNum, int idx)
        {
            if (item == (Item)null && itemNum != 0)
            {
                throw new ArgumentException("������ ���� null�Դϴ�");
            }
            _inventoryList[idx].Item = item;
            _inventoryList[idx].ItemNum = itemNum;
        }
```

honeyirene replied:
1. inventory에서 stack 직접 조작하도록 변경하기
2. if 랑 for 중괄호

honeyirene commented on Assets/Scripts/Windows/Inventory/ItemIcon.cs:
```diff
@@ -1,48 +1,89 @@
+
```
한줄 지워주세요

honeyirene commented on Assets/Scripts/Windows/Inventory/ItemListView.cs:
```diff
@@ -0,0 +1,121 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEngine;
+using UnityEngine.EventSystems;
+
+public class ItemListView : MonoBehaviour
+{
+    [SerializeField] public ItemIcon ItemIconPrefab;
```
namespace 지정 해주세요

honeyirene commented on Assets/Scripts/Windows/Inventory/ItemListView.cs:
```diff
@@ -0,0 +1,121 @@
+    }
+    public void OnOffItemIcon(bool onoff, int _index)
+    {
+        if (onoff) {
+            _ItemIconList[_index].UpdateAppearance(_inventory.TryGetItemStack(_index).Item, _inventory.TryGetItemStack(_index).ItemNum, _index);
+            _ItemIconList[_index].ItemIconOnOff(onoff);
```
중괄호

honeyirene commented on Assets/Scripts/Windows/Inventory/ItemListView.cs:
```diff
@@ -0,0 +1,121 @@
+{
+    [SerializeField] public ItemIcon ItemIconPrefab;
+    [SerializeField] public DraggingItemIcon draggingItemIcon;
+    public enum View_Category{
+        Inventory,
+        Bag,
```
중괄호

honeyirene reviewed (changes requested):
전반적으로 namespace 많이 빠져있던데 다 넣어주세요

adlet0331 committed:
PR 완료

PR

adlet0331 reviewed (commented):
PR 일차 완료

Jaewookim08 reviewed (approved):
리뷰 마무리하겠습니다.
+ namespace Inventory만 클래스 이름이랑 안 겹치게 좀 해주세요
